### ISSUE TITLE
Update the Build Environment for Ubuntu 24.04

### ISSUE THREAD

**Issue by rafaelsamenezes** [esbmc/esbmc#1003]
Since we are having demand for newer solvers and library updates. Let's update our docker image to the recent Ubuntu 23.04.

@lucasccordeiro Let me know which solvers are needed.

@fbrausse Let me know which LLVM version you are using. I would like to push for 15.

**Comment by fbrausse** [comment]
@rafaelsamenezes For CHERI I'm still on LLVM-13. Do you know whether Ubuntu 23.04 would also update cmake to 3.18 or higher?

**Comment by lucasccordeiro** [comment]
Boolector and Z3 ;-)Sent from my iPhoneOn 4 May 2023, at 11:57, Franz Brauße ***@***.***> wrote:﻿
@rafaelsamenezes For CHERI I'm still on LLVM-13. Do you know whether Ubuntu 23.04 would also update cmake to 3.18 or higher?

**Comment by rafaelsamenezes** [comment]
> Do you know whether Ubuntu 23.04 would also update cmake to 3.18 or higher?

[cmake](https://packages.ubuntu.com/lunar/devel/cmake) (3.25.1-1ubuntu1)

**Comment by fbrausse** [comment]
> @fbrausse Let me know which LLVM version you are using. I would like to push for 15.
> @rafaelsamenezes For CHERI I'm still on LLVM-13. Do you know whether Ubuntu 23.04 would also update cmake to 3.18 or higher?

Don't let my version requirement stop this, though. The Clang I'm using for CHERI is going to have to be installed separately, it's not part of the Ubuntu repository. For now, the CI will not do that and my CHERI branch should work fine with LLVM-15.

**Comment by rafaelsamenezes** [comment]
I am actually just waiting for: https://github.com/esbmc/esbmc/pull/1042

This will allow us to have more control over the environment in the future.

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes : can I ask you for the status of this PR?

**Comment by rafaelsamenezes** [comment]
> can I ask you for the status of this PR?

Still blocked by #1042, which is blocked by the CPP headers from #1131.

**Comment by fbrausse** [comment]
> I would like to push for 15.

Just noticed: Clang-14 and Clang-15 do not support _BitInt(n) for n > 128.

This affects
- regression/esbmc/bitvector_02, and
- regression/esbmc/bitvector_03

Example:
```sh
$ clang-15 -fsyntax-only ../regression/esbmc/bitvector_02/main.c
[...]
../regression/esbmc/bitvector_02/main.c:6:16: error: signed _BitInt of bit sizes greater than 128 not supported
../regression/esbmc/bitvector_02/main.c:6:16: warning: '_ExtInt' is deprecated; use '_BitInt' instead [-Wdeprecated-type]
  assert(x == (_ExtInt(1000)) y); 
               ^~~~~~~
               _BitInt
```
Before v14 and then since v16 again frontend support for those _BitInts is available in Clang. For >= v16 however, we need to solve #862 and, possibly, #1805 as well.

**Comment by fbrausse** [comment]
From https://github.com/esbmc/esbmc/pull/1581#issuecomment-1881190304:
> Starting with clang-13, the proper include paths are added for the selected GCC toolchain and manual specification is no longer necessary.

This leaves just v13 as the one with the fewest problems right now.

**Comment by fbrausse** [comment]
Update. With this resolved:
> For >= v16 however, we need to solve #862 and, possibly, https://github.com/esbmc/esbmc/issues/1805 as well.

and with users starting to depend on Clang v15+ command line options like --gcc-install-dir (see <https://github.com/esbmc/esbmc/issues/1836#issuecomment-2148259781>), I agree with @rafaelsamenezes's [comment](https://github.com/esbmc/esbmc/pull/1810#issuecomment-2102361751):
> IMO we should try to move forward into 24.04 since is the next LTS. It has from LLVM 14 up to 18.

and to go for either of the following:
1. v15: first version that supports --gcc-install-dir and last version that is fully compatible with ancient code, by which I mean the issues described in #862, or
2. v16: the _BitInt(n) with n > 128 won't be a problem anymore, but it enforces more strict -Werrors, which might be unexpected to users.

Going for 1. means least friction for users (out of the available options), while item 2. would mean a breaking change (similar to what Clang chose to do) for some users but would let us continue upgrading the Clang version in the future more easily.

@lucasccordeiro @rafaelsamenezes @mikhailramalho Do you have any opinions on this choice?

Personally, my preference would be v16: break now.

**Comment by rafaelsamenezes** [comment]
> Update. With this resolved:
> > For >= v16 however, we need to solve #862 and, possibly, https://github.com/esbmc/esbmc/issues/1805 as well.
> 
> and with users starting to depend on Clang v15+ command line options like --gcc-install-dir (see <https://github.com/esbmc/esbmc/issues/1836#issuecomment-2148259781>), I agree with @rafaelsamenezes's [comment](https://github.com/esbmc/esbmc/pull/1810#issuecomment-2102361751):
> > IMO we should try to move forward into 24.04 since is the next LTS. It has from LLVM 14 up to 18.
> 
> and to go for either of the following:
> 1. v15: first version that supports --gcc-install-dir and last version that is fully compatible with ancient code, by which I mean the issues described in #862, or
> 2. v16: the _BitInt(n) with n > 128 won't be a problem anymore, but it enforces more strict -Werrors, which might be unexpected to users.
> 
> Going for 1. means least friction for users (out of the available options), while item 2. would mean a breaking change (similar to what Clang chose to do) for some users but would let us continue upgrading the Clang version in the future more easily.
> 
> @lucasccordeiro @rafaelsamenezes @mikhailramalho Do you have any opinions on this choice?
> 
> Personally, my preference would be v16: break now.

I agree with v16

**Comment by lucasccordeiro** [comment]
Closing as done — the build environment is Ubuntu 24.04 across CI.

`ubuntu-24.04` is what the release, push, benchexec, sanitizer, build-info and symex-oracle workflows all run on, including `ubuntu-24.04-arm` for the ARM release leg and the `release-ubuntu-24.04` artifact the other workflows consume. The only older runner left is `ubuntu-22.04` in `plcplus-linux-binary.yml`, a separate PLC++ binary job rather than the main build. (The `ubuntu-20.04` strings in `pull_request.yml` are just the filename of the LLVM 11 tarball being downloaded, not a runner.)

The docker-image framing in the opening post is obsolete: there is no first-party Dockerfile and no `container:`/`image:` key in any workflow — CI runs on GitHub-hosted runners directly, so the environment is pinned per job rather than in an image.

**The Clang half of this thread is not resolved and stays open as #867.** Much of the discussion here was about which Clang to target — @fbrausse's finding that Clang 14/15 do not support `_BitInt(n)` for n > 128 (breaking `regression/esbmc/bitvector_02` and `_03`), the `--gcc-install-dir` dependency from #1836, and the conclusion that v13 had the fewest problems at the time. #867 ("Push to newer Clang") is the right home for that and is still open.
